Add support for B3 headers

ocelotl opened the pull request:
@carlosalberto The example added [here](https://github.com/lightstep/lightstep-tracer-python/pull/80/files#diff-4841a65fd27b3d6d6fe7d259e288b532) is failing because apparently I should use [this](https://github.com/opentracing/basictracer-python/blob/master/basictracer/context.py#L6) instead of [this](https://github.com/opentracing/opentracing-python/blob/master/opentracing/mocktracer/context.py#L26). I am using the latter because it (correct me if I'm wrong, please) more closely matches the OpenTracing specification for the Context, as shown [here](https://opentracing.io/docs/overview/spans/#spancontext) (the latter lacks a `sampled` attribute).

If I should use the former context class, should I use its `sampled` attribute to store the value of the very close `X-B3-Sampled` value?

carlosalberto commented on examples/http/b3_context_in_headers.py:
```diff
@@ -0,0 +1,184 @@
+import sys
+import threading
+
+try:
```
Thanks for the example. For the next iterations we will have most likely to update it, as we need to comply with a few proposed changes about how to allow the user to specify this ;)

ocelotl replied:
Sure, it is basically the same as the previous HTTP example. I can refactor it completely or create one new from scratch too.

ocelotl replied:
The example can be executed now as it is. Please let me know when we need to update it.

carlosalberto commented on lightstep/lightstep_b3_propagator.py:
```diff
@@ -0,0 +1,127 @@
+from re import compile as re_compile, IGNORECASE
+
+from opentracing.mocktracer.propagator import Propagator
+from opentracing.mocktracer.context import SpanContext
```
We **cannot** use these components as they are a **mock** component, and as such, they are not meant to be used in production code. `SpanContext` needs to be used as an interface only, and in our internal code, we extend the `basictracer` package, so we internally end up using `basictracer.SpanContext`: https://github.com/opentracing/basictracer-python/blob/master/basictracer/context.py

ocelotl replied:
Fixed!

carlosalberto commented on lightstep/lightstep_b3_propagator.py:
```diff
@@ -0,0 +1,127 @@
+                            )
+                        )
+        else:
+            if traceid is None or spanid is None:
```
We don't want to raise an `Exception` here. And here we can totally use `warn` to signal the user about this.

ocelotl replied:
Ok, will downgrade to `warn`.

ocelotl replied:
Fixed too

carlosalberto commented on lightstep/lightstep_b3_propagator.py:
```diff
@@ -0,0 +1,127 @@
+            parentspanid = carrier.get(_PARENTSPANID, None)
+            sampled = carrier.get(_SAMPLED, None)
+            flags = carrier.get(_FLAGS, None)
+            if parentspanid is sampled is flags is None:
```
`if parentspanid is sampled` doesn't look/read right. Could you confirm this is the expected check?

ocelotl replied:
Yes, I'm checking that all three of `parentspanid`, `sampled` and `flags` are `None`. I am using `is` instead of `==` because `None` [is a singleton](https://www.python.org/dev/peps/pep-0008/#programming-recommendations).

carlosalberto commented on lightstep/lightstep_b3_propagator.py:
```diff
@@ -0,0 +1,127 @@
+        )
+
+
+__all__ = ['LightStepB3Propagator']
```
We don't use this convention, so we don't need it ;)

ocelotl replied:
Ok, I'll remove it. How do you specify public interfaces, then?

carlosalberto replied:
We simply allow users to explicitly load them, i.e.

```python
from lightstep.propagators import B3Propagator
```

(This is a convention we have also kept from OpenTracing)

ocelotl replied:
Fixed

carlosalberto commented on lightstep/propagation.py:
```diff
@@ -11,10 +11,15 @@ class LightStepFormat(object):
-        tracer.inject(span.context, LightStepFormat.LIGHTSTEP_BINARY, lightstep_carrier)
+        tracer.inject(
+            span.context, LightStepFormat.LIGHTSTEP_BINARY, lightstep_carrier
+        )
```
This style change is not needed (usually, when we want to make style changes, they can come in a separated PR)

ocelotl replied:
Ok, rolling it back.

ocelotl replied:
Fixed

carlosalberto commented on lightstep/propagation.py:
```diff
@@ -11,10 +11,15 @@ class LightStepFormat(object):
     LIGHTSTEP_BINARY = 'lightstep_binary'
+    # The LIGHTSTEP_B3 format represents SpanContexts in B3 format.
+    # https://github.com/openzipkin/b3-propagation
+    LIGHTSTEP_B3 = 'lightstep_b3'
```
This can be called `B3` (`LIGHSTEP_BINARY` has this name because it's a binary specific format that only us use ;) )

ocelotl replied:
I removed it completely from this class since it is specified to be only for LightStep-specific carrier formats. The module says it is also for this kind of carrier formats. Should the B3 specification identifier declaration (`B3 = "b3"`) be on a new file?

carlosalberto commented:
Hey, I left some reviews. Let me come back to you on the sampled part (as we don't filter out based on that flag, so I need to confirm our handling in the future).

Two things:

1) It would be nice to rename `lightstep_b3_propagator.py` to `b3_propagator.py`, as B3 is not really LightStep specific ;)
2) We need tests. For some inspiration, check the ones here (in Java): https://github.com/lightstep/lightstep-tracer-java-common/blob/master/common/src/test/java/com/lightstep/tracer/shared/B3PropagatorTest.java

ocelotl commented on lightstep/lightstep_b3_propagator.py:
```diff
@@ -0,0 +1,127 @@
+            sampled = baggage.get(self._SAMPLED, None)
+            if sampled is not None:
+                if flags == 1:
+                    warn(
```
Ok, changing to logger.

ocelotl replied:
Fixed!

carlosalberto commented on lightstep/b3_propagator.py:
```diff
@@ -0,0 +1,136 @@
+            if baggage == OTSpanContext.EMPTY_BAGGAGE:
+                baggage = None
+        return SpanContext(
+            # traceid and spanid are encoded in hex, so thet must be encoded
```
If no valid `traceid`/`spanid` were encoded, we need to report it as done here: https://github.com/opentracing/basictracer-python/blob/master/basictracer/text_propagator.py#L51

ocelotl replied:
Done, raising `SpanContextCorruptedException`.

carlosalberto commented on lightstep/propagation.py:
```diff
@@ -4,6 +4,10 @@
 
+# The B3 format represents SpanContexts in B3 format.
+# https://github.com/openzipkin/b3-propagation
+B3 = "b3"
```
Do we really need it here?

ocelotl replied:
Well, it can be placed somewhere else if you prefer, of course. I just added it there since that module was holding the register identifier for the other propagator in the package, `LightStepBinaryPropagator`.

carlosalberto commented on lightstep/tracer.py:
```diff
@@ -75,6 +76,7 @@ def __init__(self, enable_binary_format, recorder, scope_manager):
         super(_LightstepTracer, self).__init__(recorder, scope_manager=scope_manager)
         self.register_propagator(Format.TEXT_MAP, TextPropagator())
         self.register_propagator(Format.HTTP_HEADERS, TextPropagator())
+        self.register_propagator(B3, B3Propagator())
```
Let's leave this as an optional step (so by default only `TextMapPropagator` is registered by default).

ocelotl replied:
Done, removed `B3` as a default propagator.

ocelotl commented:
@carlosalberto I added several more test cases, and fixed all but [one of your comments](https://github.com/lightstep/lightstep-tracer-python/pull/80/files/04eaa423ef44b81952555bd59d0a1ff5c70bff02#diff-4c97a3e758defaa6a9f418bdfab75551) since I need a bit more info still.

Thank you!

chadbrad commented:
@carlosalberto and / or @codeboten - are we good to merge?

carlosalberto commented:
Hey @ocelotl 

Thanks. Actually we need a last iteration, which I hadn't realize. The way injection/extraction works is to specify one of the formats in `opentracing.Format`, and not 'create' a new one to use B3 (or later `TraceContext`), **but** use B3 when either `TEXT_MAP` or `HTTP_HEADERS` are specified:

```python
# replaces the default propagator with the B3 one
tracer.register_propagator(b3_propagator)
...
# we use the text map format, with the b3 propagator
tracer.inject(span.context, opentracing.Format.TEXT_MAP, carrier)
```

For this you will need to also update the example ;) Once this is done, we will be able to merge.

ocelotl commented:
@codeboten @carlosalberto I implemented the truncation of the MSb in `traceid` in the output of `extract` until `traceid` is 64b long. I still added `FIXME` comments to highlight the pending question of padding/truncating on the MS or LS side.

ocelotl commented:
> @codeboten @carlosalberto I implemented the truncation of the MSb in `traceid` in the output of `extract` until `traceid` is 64b long. I still added `FIXME` comments to highlight the pending question of padding/truncating on the MS or LS side. I'll squash all this into a single before merging if you prefer.



> Hey @ocelotl
> 
> Thanks. Actually we need a last iteration, which I hadn't realize. The way injection/extraction works is to specify one of the formats in `opentracing.Format`, and not 'create' a new one to use B3 (or later `TraceContext`), **but** use B3 when either `TEXT_MAP` or `HTTP_HEADERS` are specified:
> 
> ```python
> # replaces the default propagator with the B3 one
> tracer.register_propagator(b3_propagator)
> ...
> # we use the text map format, with the b3 propagator
> tracer.inject(span.context, opentracing.Format.TEXT_MAP, carrier)
> ```
> 
> For this you will need to also update the example ;) Once this is done, we will be able to merge.

Ok! implementing...

ocelotl commented:
> > @codeboten @carlosalberto I implemented the truncation of the MSb in `traceid` in the output of `extract` until `traceid` is 64b long. I still added `FIXME` comments to highlight the pending question of padding/truncating on the MS or LS side. I'll squash all this into a single before merging if you prefer.
> 
> > Hey @ocelotl
> > Thanks. Actually we need a last iteration, which I hadn't realize. The way injection/extraction works is to specify one of the formats in `opentracing.Format`, and not 'create' a new one to use B3 (or later `TraceContext`), **but** use B3 when either `TEXT_MAP` or `HTTP_HEADERS` are specified:
> > ```python
> > # replaces the default propagator with the B3 one
> > tracer.register_propagator(b3_propagator)
> > ...
> > # we use the text map format, with the b3 propagator
> > tracer.inject(span.context, opentracing.Format.TEXT_MAP, carrier)
> > ```
> > 
> > 
> > For this you will need to also update the example ;) Once this is done, we will be able to merge.
> 
> Ok! implementing...

@carlosalberto Done!

codeboten commented on lightstep/b3_propagator.py:
```diff
@@ -0,0 +1,149 @@
+        carrier.update(baggage)
+
+        if traceid is not None:
+            # FIXME This adds zeroes on the MS side until traceid is 128b long.
```
padding must be on the least significant bits

codeboten commented on lightstep/b3_propagator.py:
```diff
@@ -0,0 +1,149 @@
+            if baggage == OTSpanContext.EMPTY_BAGGAGE:
+                baggage = None
+
+        # FIXME This truncates the MSb of traceid until only 64b are left.
```
the implementation here should leave the trace ID as is when it's propagated to ensure it matches what was passed in. the truncation should only occur when reporting to lightstep

ocelotl replied:
Sure, I'll change it. What do you refer to with `when reporting to lightstep`?

codeboten replied:
A good example of this is in the ruby implementation, where a separate field stores the 128-bit vs the 64-bit trace IDs, to remain compatible with the current propagators: https://github.com/lightstep/lightstep-tracer-ruby/pull/86/files#diff-d87db7512b8e96c0a8d9463d52de94beR13-R14. This ensures existing propagation continues to work as is.

ocelotl commented:
@codeboten Changes added, please confirm :+1:

codeboten commented on lightstep/tracer.py:
```diff
@@ -73,8 +73,8 @@ class _LightstepTracer(BasicTracer):
-        self.register_propagator(Format.TEXT_MAP, TextPropagator())
-        self.register_propagator(Format.HTTP_HEADERS, TextPropagator())
+        self.register_propagator(Format.TEXT_MAP, B3Propagator())
+        self.register_propagator(Format.HTTP_HEADERS, B3Propagator())
```
Think we wanted the default here to continue being the TextPropagator

ocelotl replied:
This was requested by @carlosalberto before:

> Hey @ocelotl
> 
> Thanks. Actually we need a last iteration, which I hadn't realize. The way injection/extraction works is to specify one of the formats in `opentracing.Format`, and not 'create' a new one to use B3 (or later `TraceContext`), **but** use B3 when either `TEXT_MAP` or `HTTP_HEADERS` are specified:
> 
> ```python
> # replaces the default propagator with the B3 one
> tracer.register_propagator(b3_propagator)
> ...
> # we use the text map format, with the b3 propagator
> tracer.inject(span.context, opentracing.Format.TEXT_MAP, carrier)
> ```
> 
> For this you will need to also update the example ;) Once this is done, we will be able to merge.

codeboten replied:
Ah, my understanding is that both TEXT_MAP and HTTP_HEADERS should default to TextPropagator, unless configured otherwise when configuring the Tracer.

ocelotl replied:
@carlosalberto can you confirm, please?

carlosalberto replied:
Yes, we want `TextPropagator` being the default.

> not 'create' a new one to use B3 (or later TraceContext), but use B3 when either TEXT_MAP or HTTP_HEADERS are specified

Correct, we want `B3` to **completely** replace `TextPropagator` **when** registered explicitly (instead of having it as a separate format).

ocelotl replied:
I [reverted](https://github.com/lightstep/lightstep-tracer-python/blob/70da71a1eb75b32ba831f3e3ba9295e368521a87/lightstep/tracer.py#L76-L77) the changes in `tracer.py`. I also [updated](https://github.com/lightstep/lightstep-tracer-python/blob/70da71a1eb75b32ba831f3e3ba9295e368521a87/examples/http/context_in_headers.py#L151-L154) the example to explicitly call `register_propagator` twice to register the B3 propagator for `TEXT_MAP` and `HTTP_HEADERS` respectively.

Just to be sure, @carlosalberto, you mentioned this in your previous comment:
```
# replaces the default propagator with the B3 one
tracer.register_propagator(b3_propagator)
```
But is not possible to call `register_propagator` with only one argument, the format name and propagator instance are needed.

Please let me know if this is the right implementation.

Thanks!

carlosalberto replied:
> But is not possible to call register_propagator with only one argument, the format name and propagator instance are needed.

Yeah, this is fine ;)

codeboten commented:
LGTM 👍
